pkg/tfvars/tfvars.go

```diff
@@ -23,102 +21,62 @@ type config struct {
 	IgnitionBootstrap string `json:"ignition_bootstrap,omitempty"`
 	IgnitionMaster    string `json:"ignition_master,omitempty"`
 
-	aws.AWS             `json:",inline"`
-	libvirt.Libvirt     `json:",inline"`
-	openstack.OpenStack `json:",inline"`
+	*aws.AWS             `json:",inline"`
+	*libvirt.Libvirt     `json:",inline"`
+	*openstack.OpenStack `json:",inline"`
 }
 
-// TFVars converts the InstallConfig and Ignition content to
-// terraform.tfvar JSON.
-func TFVars(cfg *types.InstallConfig, bootstrapIgn, masterIgn string) ([]byte, error) {
+// TFVars generates terraform.tfvar JSON for launching the cluster.
+func TFVars(clusterID string, clusterName string, baseDomain string, vpcCIDR string, bridge string, image string, cloud string, bootstrapIgn string, masterIgn string, masters []*clusterapi.Machine) ([]byte, error) {
 	config := &config{
-		ClusterID:  cfg.ClusterID,
-		Name:       cfg.ObjectMeta.Name,
-		BaseDomain: cfg.BaseDomain,
-
-		IgnitionMaster:    masterIgn,
+		ClusterID:         clusterID,
+		Name:              clusterName,
+		BaseDomain:        baseDomain,
+		Masters:           len(masters),
 		IgnitionBootstrap: bootstrapIgn,
+		IgnitionMaster:    masterIgn,
 	}
 
-	for _, m := range cfg.Machines {
-		switch m.Name {
-		case "master":
-			var replicas int
-			if m.Replicas == nil {
-				replicas = 1
-			} else {
-				replicas = int(*m.Replicas)
-			}
+	if config.Masters == 0 {
+		return nil, errors.Errorf("masters slice cannot be empty")
+	}
 
-			config.Masters += replicas
-			if m.Platform.AWS != nil {
-				config.AWS.Master = aws.Master{
-					EC2Type:     m.Platform.AWS.InstanceType,
-					IAMRoleName: m.Platform.AWS.IAMRoleName,
-					MasterRootVolume: aws.MasterRootVolume{
-						IOPS: m.Platform.AWS.EC2RootVolume.IOPS,
-						Size: m.Platform.AWS.EC2RootVolume.Size,
-						Type: m.Platform.AWS.EC2RootVolume.Type,
-					},
-				}
-			}
-		case "worker":
-			if m.Platform.AWS != nil {
-				config.AWS.Worker = aws.Worker{
-					IAMRoleName: m.Platform.AWS.IAMRoleName,
-				}
-			}
-		default:
-			return nil, errors.Errorf("unrecognized machine pool %q", m.Name)
-		}
+	machine := masters[0]
```

Comment:
This entire section should not use this `map[string]interface{}`. each provider should have schema.

Reply:
> This entire section should not use this `map[string]interface{}`. each provider should have schema.

The input master `*Machine` instances already specify the platform via the provider's `apiVersion` field (e.g. `aws.cluster.k8s.io/v1alpha1`), so switching on that seemed like a DRY way to determine which platform to use to translate the `Machine` content into Terraform.  It also allows me to push the platform-specific provider unpacking down into the provider-specific subpackages.  What alternative platform identifier would you like me to switch on instead?

Or are you just saying you'd rather I unmarshal into a:

```go
type apiVersionWrapper struct {
  APIVersion `json:"apiVersion"`
}
```

instead of doing the [arbitrary-data decode dance][1]?  Or something else?

[1]: https://blog.golang.org/json-and-go#TOC_5.

Reply:
if the machine is a type, then you can directly switch on `platform.(type)`

Reply:
> if the machine is a type, then you can directly switch on `platform.(type)`

Done in 2b2e8c149e5.